VerilogCodingStyle.md

```diff
@@ -1797,92 +1797,171 @@ increment.
 
 ### Don't Cares (`X`'s)
 
-***Explicitly specify don't cares when safe to do so. Don't silently
-squash `X` in logic.***
+***The use of `X` literals in RTL code is strongly discouraged. RTL must not
+assert `X` to indicate "don't care" to synthesis in any case.***
 
-Don't Care values can significantly help the quality of logic optimization and
-should be used wherever it is safe to do so. Explicitly declaring unused decodes
-as `X` has the secondary effect of propagating `X`'s through the design if the
-**input unused** design assumption is violated, making those bugs easier to
-diagnose and fix.
+If not strictly controlled, the use of `X` assignments in RTL to flag invalid or
+don't care conditions can lead to simulation/synthesis mismatches.
 
-Example:
+Instead of assigning and propagating `X` in order to flag and detect invalid
+conditions, it is encouraged to make **extensive use of SVAs**. The added
+benefits of this design practice is that
 
-```systemverilog
-always_comb begin
-  unique case (state)
-    ALPHA:   decode = 16'd1;
-    BETA:    decode = 16'd127;
-    GAMMA:   decode = 16'd43;
-    // all other states are unused:
-    default: decode = 'X;  // or {16{1'bx}} in Verilog-2001
-  endcase
-end
-```
+- No special code style is required to properly propagate `X` conditions,
+- The chance of accidentally introducing simulation/synthesis mismatches is
+  systematically reduced,
+- Simulation breaks early and less signal backtracking is needed to
+  root-cause bugs,
+- In several cases, formal property verification (FPV) can be used to prove
+  whether these SVAs can always be fulfilled.
 
-Write logic that will either propagate `X` or assert when your inputs are `X`.
-Avoid silently squashing `X` by writing logic that resolves an `X` input to a
-non-`X` output. Instead, write logic that either explicitly propagates the `X`
-or uses an assert to raise an exception when an input is `X`.
+The solution presented here has similarities with the approaches presented in
+["Being Assertive With Your X"](http://www.lcdm-eng.com/papers/snug04_assertiveX.pdf)
+by Don Mills.
 
-Be aware: Any logical operation involving `X` always propagates the `X`.
+Note that although don't cares can be used to indicate possible optimization
+opportunities to the synthesis tool, it is debatable whether the gains in logic
+reduction are significant enough to outweigh the possible synthesis mismatch
+issues that the use of `X` literals may entail (especially with the gate-counts
+available in today's technologies). Large LUTs may be an exception to this
+statement and the use of `X` in that context should be carefully evaluated on a
+case-by-case basis.
```

Comment:
Yeah I completely agree with your point, although I think that should probably go into our additional RTL coding guide for security critical applications.

Reply:
WDYT, @sjgitty ?

Reply:
If by large LUTs you mean large roms encoded with case statements or similar I don't think it's unreasonable to say you must use 0 or 1 for the bits you don't care about (designers choice, e.g. if you want to output 8'b1xxxxxxx then 1 may be the sensible choice as it matches the other bit or if there's lot of inputs that are total don't cares on output then try to match the most used bit say you output either one or two 1s with 0s and otherwise Xs the don't care cases would become zeros).

You could also say that if you end up with a 'sparse LUT' (i.e. one where we don't care about much of the inputs and/or outputs) then you may be doing it wrong and should rethink.

As a final fallback recommend you push the LUT with Xs through some logic minimization tool to get the same LUT back but with choices for the Xs that work well (though I'm not sure such a tool exists you'll just get a bunch of gates and we may get back to the same information leakage issue).

Reply:
Indeed, I meant variations of large case statements or arrays.

In the past, people have been using tools like [Espresso](https://en.wikipedia.org/wiki/Espresso_heuristic_logic_minimizer) to minimize large truth tables - although in my own experiments I got the best results by specifying large LUTs with arrays in SystemVerilog and running them straight through DC. 

I never really specified sparse LUTs in this manner, but I did use truncated arrays for unaligned LUTs. I.e., say you have an LUT with 567 words, I would use an array `[0:566]`, which implicitly means that `[567:1023]` are `X` when indexed with higher indexes. But then again padding with zero up to the next pow2 depth, or using a guarding if statement around the indexing operation would likely not add much more logic and would make this predictable and safe.

So I would say, let's make this strict and just remove that statement on LUTs.